File: Binary Tree Maximum Path Sum.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 class Solution {
private:
    int maxSumEndAt(TreeNode* root, int& result) {
        if (!root)
            return 0;
		int left = max(0, maxSumEndAt(root->left, result));
		int right = max(0, maxSumEndAt(root->right, result));
        int sum = root->val + max(left, right);
        if (root->val + left + right > result)
            result = root->val + left + right;
        return sum;
    }
public:
    int maxPathSum(TreeNode *root) {
        int result = INT_MIN;
        maxSumEndAt(root, result);
        return result;  
    }
};

File: Binary Tree Maximum Path Sum.py
# Definition for a  binary tree node
# class TreeNode:
#	 def __init__(self, x):
#		 self.val = x
#		 self.left = None
#		 self.right = None

class Solution:
	def maxSumEndAt(self, root, result):
		if (root == None):
			return 0
		left = max(0, self.maxSumEndAt(root.left, result))
		right = max(0, self.maxSumEndAt(root.right, result))
		sumThroughRoot = root.val + left + right
		if (sumThroughRoot > result[0]):
			result[0] = sumThroughRoot
		return root.val + max(left, right)
	# @param root, a tree node
	# @return an integer
	def maxPathSum(self, root):
		if (not root):
			return 0
		result = [root.val]
		self.maxSumEndAt(root, result)
		return result[0]
		